synfig-studio/src/gui/states/state_bone.cpp

```diff
@@ -731,148 +737,17 @@ StateBone_Context::event_mouse_release_handler(const Smach::event& x)
-							Point aOrigin = matrix.get_transformed(clickOrigin);
-							aOrigin[0]/=sx;
-
-							createChild->set_param("value_desc",Action::Param(v_d));
-							createChild->set_param("origin",Action::Param(ValueBase(aOrigin)));
-							createChild->set_param("width",Action::Param(ValueBase(get_bone_width())));
-							createChild->set_param("tipwidth",Action::Param(ValueBase(get_bone_width())));
-							createChild->set_param("prev_active_bone_node",Action::Param(get_work_area()->get_active_bone_value_node()));
-							if((clickOrigin-releaseOrigin).mag()>=0.01) {
-								a = atan2((releaseOrigin-clickOrigin)[1],(releaseOrigin-clickOrigin)[0]);
-								createChild->set_param("angle",Action::Param(ValueBase(Angle::rad(a-angle))));
-								createChild->set_param("scalelx", Action::Param(ValueBase((releaseOrigin - clickOrigin).mag())));
-							}else{
-								createChild->set_param("angle",Action::Param(ValueBase(Angle::rad(a-angle))));
-								createChild->set_param("scalelx", Action::Param(ValueBase(1.0)));
-							}
-							if(createChild->is_ready()){
-								try{
-									get_canvas_interface()->get_instance()->perform_action(createChild);
-									value_desc= ValueDesc(list_node,active_bone,list_desc);
-
-								} catch (...) {
-									info("Error performing action");
-								}
-							}
-						}else{
-							Action::PassiveGrouper group(get_canvas_interface()->get_instance().get(),"Add Bone");
-
-							Action::Handle action = Action::create("ValueNodeStaticListInsert");
-							action->set_param("canvas", get_canvas());
-							action->set_param("canvas_interface", get_canvas_interface());
-							action->set_param("time", get_canvas()->get_time());
-
-							Bone bone1 = Bone();
-
-							bone1.set_parent(ValueNode_Bone_Root::create(Bone()));
-							bone1.set_origin(clickOrigin);
-							bone1.set_width(get_bone_width());
-							bone1.set_tipwidth(get_bone_width());
-							bone1.set_angle(Angle::rad(acos(0.0)));
-							if((clickOrigin-releaseOrigin).mag()>=0.01) {
-								bone1.set_scalelx((releaseOrigin - clickOrigin).mag());
-								bone1.set_angle((releaseOrigin - clickOrigin).angle());
-							}
-
-							Bone bone2 = Bone();
-
-							bone2.set_parent(ValueNode_Bone_Root::create(Bone()));
-							bone2.set_origin(clickOrigin);
-							bone2.set_width(get_bone_width());
-							bone2.set_tipwidth(get_bone_width());
-							bone2.set_angle(Angle::rad(acos(0.0)));
-							if((clickOrigin-releaseOrigin).mag()>=0.01) {
-								bone2.set_scalelx((releaseOrigin - clickOrigin).mag());
-								bone2.set_angle((releaseOrigin - clickOrigin).angle());
+							ValueNode::Handle new_active_bone_node;
+							if (skel_layer) {
+								new_active_bone_node = ValueNode_Bone::create(bone,get_canvas());
+								action->set_param("item",new_active_bone_node);
+							} else {
+								ValueNode_Composite::Handle bone_pair = ValueNode_Composite::create(std::pair<Bone,Bone>(bone,bone),get_canvas());
```

Comment:
```
std::pair<Bone,Bone>(bone,bone)
```
I am not sure - is it safe to use a same single `bone` to create a bone pair here?
Maybe use two bone objects (`bone1` and `bone2`)?

Reply:
I rechecked now and it's safe.
It'll actually store a value copy.

For this case, `ValueNode_Composite::create` calls `ValueNode_Const::create` with a `Pair<Bone,Bone>` value, and that calls `ValueNode_Bone::create` that _finally_ calls `ValueNode_Bone::ValueNode_Bone(const ValueBase &value, etl::loose_handle<Canvas> canvas)` for each Pair<Bone,Bone> item component (first and second).

In this last method, it uses the pair item component (first or second) here (named as `value`): `Bone bone(value.get(Bone()));`

Reply:
Got it! Thank you for the explanation!